### ISSUE TITLE
Question: How would I go about making a persistent player....

### ISSUE THREAD

**Issue by rchrdnsh** [E-Kuerschner/useAudioPlayer#6]
Hi again,

Sorry if I'm blowing your issues up, but this is a really nice package, and I'm curious if it might be able to handle my use case.

I have made this app:

rykr.netlify.com

Which is a blog and music player in one, with a persistent player on the bottom of the screen at all times. I am trying to run all audio though it, but I currently need to define a set of files for the player to play at one time, like so:

```jsx
const { tracks, assets } = useStaticQuery(graphql`
    query Tracks {
      tracks: allMdx(filter: { fileAbsolutePath: { regex: "/content/music/" } }
      ) {
        edges {
          node {
            fields {
              slug
            }
            frontmatter {
              name
              artist
              genre
              bpm
              # ADD BASE HERE
              artwork {
                base
                childImageSharp {
                  fluid(maxWidth: 1000) {
                    ...GatsbyImageSharpFluid
                  }
                }
              }
              alt
              description
              release(formatString: "MMMM Do, YYYY")
              audio {
                absolutePath
                base
              }
            }
          }
        }
      }

      # query all mp3 and jpg files from /content/music/
      assets: allFile(
        filter: {
          extension: { in: ["mp3", "jpg"] }
          absolutePath: { regex: "/content/music/" }
        }
      ) {
        edges {
          node {
            publicURL
            relativePath
          }
        }
      }
    }
  `)

  // This uses the map function to create an array of objects, where each object is a track and all it's information and assets. We need to use `useMemo` to avoid re-computation when the state changes as this information is all static.
  const trackList = useMemo(
    () =>
      tracks.edges.map(track => {
        const slug = track.node.fields.slug
        const { frontmatter } = track.node
        const {
          name,
          artist,
          genre,
          bpm,
          artwork,
          alt,
          description,
          audio,
        } = frontmatter

        return { slug, name, artist, genre, bpm, artwork, alt, description, audio }
      }),
    [tracks]
  )
```

...so I am using a graphql `staticQuery` in gatsby to generate an array of tracks, which I then pull in to the player and it works ok, like so:

```jsx
 const [state, setState] = useState({
    audioPlayer: {},
    currentTrackIndex: null,
    isPlaying: false,
  })
  
  const [currentTime, setCurrentTime] = useState(state.audioPlayer.currentTime)
  
  useEffect(() => {
    setState({
      audioPlayer: new Audio(),
      currentTrackIndex: null,
      isPlaying: false,
    })
  }, [])
  
  const formattedTime = getTime(currentTime) || `0:00`
  const formattedDuration = getTime(state.audioPlayer.duration) || `0:00`

// etc, etc etc...
```

...but what I really want is to have pieces of audio on many different pages in the app, and all of the audio everywhere is connected to the audio context and can be played, and when it is played it is then added to the array, and this would include full songs, audio snippets, sound effects, examples, etc...all on different pages, but all being sent through one audio context.

I have no idea who to go about doing this, but was curious if you might have any ideas about how this might be possible to do in a react and/or gatsby app, or even just in general.

This is more a discussion, I guess, then any specific issue, just want to get your thoughts, if you have the time and interest to share them :-)

**Comment by E-Kuerschner** [comment]
> Sorry if I'm blowing your issues up, but this is a really nice package, and I'm curious if it might be able to handle my use case.

This is what open source is for is it not? 🙃 

One point of clarification: when you say the audio gets added to an array, is this array intended to be playlist?

But in general to get started I think you would need to place the AudioPlayerProvider way up near the root of your application (or at least as high as the common ancestor to all components that need to initial playing audio through the single audio context). For example, if you application is using a route for different pages (each page triggering audio) then you will want to make sure to wrap your Router in the provider.
```
const App = () => {
  <AudioPlayerProvider>
    <Router>
      {/* routes to different pages */}
    </Router>
  </AudioPlayerProvider>
}
```
Then, in any of your route components or their children you could use the `useAudioPlayer` hook to manipulate this one audio context at the root of your application. In order to get the playlist functionality you may have to wrap the AudioPlayerProvider to enhance it with this functionality since this library currently does not support queuing up audio to play when the current one ends. Although, i think that is a feature worth considering.

**Comment by rchrdnsh** [comment]
Yeah, sorry, didn't clarify :-( 

I already have the app wrapped in an `AudioContext` of my own design, like so:

```jsx
const AudioContext = React.createContext([{}, () => {}])

const AudioProvider = props => {

// .....code, so much code.....

// Creating a hook to access global audio state from this Audio Context file
function useAudioState() {
  return useContext(AudioContext)
}

export { useAudioState, AudioProvider }

// then, in my layout file(how gatsby thinks of the app file)...

return (
    <AudioProvider>
      <App>
        <Logo to={`/`}>
         // ...code...
        </Logo>
        <Menu>
         // ...code...
        </Menu>
        <MDXProvider components={{ ...components }}>
          <Main>
            {children}
          </Main>
        </MDXProvider>
        <Player>
          <AudioPlayer/>
        </Player>
      </App>
    </AudioProvider>
  )
}
```

...so that part is good.

And yes, basically it would be like a play history, I suppose, although that part isn't even as important as simply having the ability to have a persistent set of playback controls, and then any audio on any page would be piped into those controls, ala spotify, or any of streaming service for that matter, and take over for the previously playing piece of music or audio clip or whatever.

Currently in my `AudioContext` file I query all the audio for the music, although I would much rather not do that and simply have a way of wrapping each piece of audio in some component that tapped into the global audio context, so that I could wrap literally anything that is audio anywhere on any page and it would play through the global audio playback controls...I think `useAudioplayer` is probably the thing that I'm looking for, although how to use it properly, I do not yet know.

I also don't have any pre-loading or caching or chunking or sprite-ing or HLS'ing or even a loading state UI so users currently hit the play button on a track and have to wait for 10 or more seconds for the audio to play, even though the time is ticking away, which is a bad UX for an app whose main purpose it is to stream audio well.

hence, why I found this package of yours :-)

**Comment by E-Kuerschner** [comment]
So rather than querying all the audio upfront inside the AudioContext, you could distribute that responsibility to the different pages as you are hoping you can. The AudioContext will still be the component actually doing the fetching but it won't fetch until told to do so by someone taking action on one of those pages. 

`useAudioPlayer` accepts an argument which is the path to load audio from. When this changes it will load the new sound and begin playing. I currently don't support lazy loading the audio or queuing up sounds to play one after the other but I think I could start working on those features next. It seems like a natural progression. 

In the mean time I have pushed up a new example where we can start to experiment with this type of setup.

**Comment by rchrdnsh** [comment]
hmmmm...that does make some sense...and thank you for your help :-) Gonna check out the example and see if I can wrap my head around it. I am also relatively new to programming and react, so I have holes in my knowledge as well, which does not help, but as I dive deeper into trying to build things that interest me, I'm able to unearth new holes in my knowledge, which is really helpful. So if I ever ask any questions that may seem simple, it's because of these gaps in my understanding that I sometimes need a little help seeing how things work :-)

**Comment by rchrdnsh** [comment]
ok, chill example :-)

I added an extra play/pause button to each page by replicating the PlayBar button for the dog and cat pages...but I'm noticing a UX issue. If I hit play on the dog page, then switch to the cat page, the cat sound starts playing automatically, taking over from the dog sound, but what should happen is that the dog sound keeps playing when i switch to the cat page, and only when i click play on the cat sound, should it take over...do you think would be easy to achieve with your package, or is that something that would be more difficult to pull off?

**Comment by E-Kuerschner** [comment]
Yeah that is expected behavior given the way it is currently implemented. What is happening is when you visit the cat page the `Cat` component is rendered for the first time and this line of code runs:
`useAudioPlayer('cats.mp3')`
This causes the AudioPlayerContext to immediately load this file (and since it was previously playing a sound it will autoplay once it finishes loading).

What you want is a way to lazily load the audio at a later point in time. I do plan on building that in to the API so that it is a little more easier to achieve but for now it should be possible by setting things up a little differently:

Basically you would have to have a component to manage loading audio files by changing src config in `useAudioPlayer`.

```
const AudioLoader = () => {
  const [file, setFile] = React.useState(null)
  // im not sure if it supports null so you may have to have some default until that is fixed
  useAudioPlayer({ src: file })
  // you call this in you app when you want to change tracks
  const changeTrack = (fileName) => {
    setFile(fileName)
  }

  ... however you want to go forward: either render the rest of your app here or use context to pass 
  changeTrack down to chidren that need it
}
```

**Comment by rchrdnsh** [comment]
i will give this a try...and yeah, if you can build this package out to support a basic Spotify style app interface with all the nice little trimmings, that would be amazing, indeed :-)

**Comment by E-Kuerschner** [comment]
Working on a queuing feature today. Closing this out in the meantime.

**Comment by E-Kuerschner** [comment]
@rchrdnsh released a new version 0.0.11 which supports lazy loading. Check out the GlobalPlayer example now. I think this should be a little closer to what you wanted

**Comment by rchrdnsh** [comment]
hmmmm...not exactly what I was thinking, although it seems to work as you expect it to :-)

It seems as if the audio files don't load until the `Let's hear 'em` button is pressed, but then the user would still have to hit the play button on the bottom to make it work, is this supposed to be correct?

What I was expecting was for the the audio to load instantly when switching to another page, but then wait to be played until the play button next to the audio sample is clicked, which would then take over the global play button, does that make sense? So when navigating to a new page, all of the audio samples load, but are not triggered until a user gesture, which would be a play button for each audio clip/song/track/etc...

Sort of like this:

Imagine a grid of albums:

<img width="1426" alt="Screen Shot 2020-01-19 at 10 54 56 PM" src="https://user-images.githubusercontent.com/7884263/72704859-b924d080-3b0e-11ea-8d54-a772ccf749c4.png">

...where each album has a play button that appears when you hover over it, or the play button is always visible. When the page loads, all of the audio samples on the page should begin loading, maybe in order? or not, whatever, and then when the user clicks on one of the play buttons then the song starts playing and all of the info shows up in the global player controls at the bottom of the page. Now, if the user navigates to a different page, then the music keeps playing in the global player, but all the new music on the new page starts loading, and then when the user clicks the play button on any song in the new page then that song would take over the global player controls, but the previous song would have kept playing until the new song is started.

I may be over-explaining, I'm not sure, but yeah, that's more along the lines of what I was thinking. Lazy-loading is definitely a nice feature, just not one I can imagine using much, as most of the time the user would want all audio to instantly play, no?

The site I'm working on is an another example, where every song has a play button, but only one is active at a time, like so:

<img width="1434" alt="Screen Shot 2020-01-19 at 11 03 33 PM" src="https://user-images.githubusercontent.com/7884263/72705333-0e151680-3b10-11ea-81aa-77846e602ff7.png">

...but I have rolled this mostly by myself with a massive amount of help from @universse, who is an absolute boss, by the way, but it still has quite a few issues that are a bit beyond me, hence seeking out a library like howler, which then led me here, to your awesome work :-)

I guess ultimately what I'm looking for is the ability to add a `Play Button` component to any piece of audio and have it tap in to the global audio context, ala, Spotify, YouTube Music, Apple Music, Tidal, etc, then have instant loading, or a lazy-loading option if I need it?

**Comment by E-Kuerschner** [comment]
Cool, I'm tracking what you are saying. I will publish a roadmap this week detailing the features I plan on rolling out. Supporting multiple audio sources being loaded at once is currently not supported. Only one audio source can be loaded at a time. This is why I load the new audio ("let's hear em" button in the GlobalPlayer example) when the user wants to start listening to it instead of what is already playing. Luckily the time to load the new audio in this setup is pretty quick - almost instantaneous, but in a more realistic environment there would probably be a long enough delay to hurt the user's experience who wants to listen to their song the moment they click the mouse.

**Comment by rchrdnsh** [comment]
awesome. Will chech it out out and chime in as well :-)
